Build experimental branches with prefix for public review

aj-stein-nist opened the pull request:
Update scripts and workflows and close #19.

This will update the workflow to allow the following:

1. Override the revisions and not compute them.
1. Override the prefix and ignore overridden revisions.
1. The default, add prototype* as a prefix for publication from the
usnistgov/OSCAL repo.

In the case the last scenario is not desired, a developer with admin
permissions in the repository can provide a '' prefix and all prototype
prefix branches go away and will be silently ignored.

aj-stein-nist commented:
@nikitawootten-nist, I am not sure we would want to or not want to adjust the cache keys based on tags as opposed to accompanying revisions, that would make this maybe more complicated, or do I just have to find a way of calling the [updated block in the relevant GitHub Actions block](https://github.com/usnistgov/OSCAL-Reference/blob/d5eac0ba934c301cc537f35d7c5065ffcea83d5e/.github/workflows/pages.yaml#L32)?

nikitawootten-nist commented on support/list_revisions.sh:
```diff
@@ -15,4 +15,20 @@ TAGGED_REVISIONS=$(echo "${TAGS}" | grep -E '^v[0-9]+\.[0-9]+\.[0-9]+$')
 TAGGED_REVISIONS=$(echo "${TAGGED_REVISIONS}" | sort -t "." -k1,1n -k2,2n -k3,3n)
 
-echo "${TAGGED_REVISIONS}"
+# retrieve prefix for branches of prototypes to be listed and published and if
+# it is empty, it can be ignored.
+PROTOTYPE_BRANCHES_REMOTE=${PROTOTYPE_BRANCHES_REMOTE:-""}
+PROTOTYPE_BRANCHES_PREFIX=${PROTOTYPE_BRANCHES_PREFIX:-""}
+
+if [[ -z "${PROTOTYPE_BRANCHES_PREFIX}" || -z "${PROTOTYPE_BRANCHES_REMOTE}" ]]; then
+    # the default values are empty, do nothing
+    echo "${TAGGED_REVISIONS}"
+else
+    # the output of ls-remote is the best for this but doesn't have ready-made
+    # format or pretty-print options. We need to use sed to filter the output
+    # output is the following format:
+    # sha1-hash        refs/heads/branch-name
+    # 012345abcd012345abcd012345abcd012345abcd        refs/heads/branch-name
+    PROTOTYPE_BRANCHES=$(cd "${OSCAL_DIR}"; git ls-remote "${PROTOTYPE_BRANCHES_REMOTE}" "${PROTOTYPE_BRANCHES_PREFIX}*" | sed -n -e "s|^.*\(refs/heads/\)\(${PROTOTYPE_BRANCHES_PREFIX}\)|\2|p")
+    echo "${TAGGED_REVISIONS}" "${PROTOTYPE_BRANCHES}"
+fi
```
B: Scripts such as the release asset generator and modeldoc generator rely on the latest release being last (so that it can be extracted with a simple `./list_revisions.sh | tail -n 1`). The script in its current state will break that behavior.

I would request that for consistency we put the prototype revision extraction in a separate script (or directly in the Makefile). This is why for example the Makefile appends `develop`, as the `list_revisions.sh` script's only concern is extracting tagged versions.

nikitawootten-nist replied:
Maybe part of this PR could be communicating that more clearly, renaming the script to `list_tagged_releases.sh` or something of that sort.

aj-stein-nist replied:
So that field appends and not prepends `develop`? OK, let's talk on Gitter about how we order the script invocations. I was not aware of the order requirement, I'm sorry. 🤦

nikitawootten-nist replied:
My mistake, the Makefile prepends develop to the `REVISIONS` variable, but both of the generation scripts rely on `list_revisions.sh` to get taggged releases. The order of the `REVISIONS` variable in the Makefile does not matter, but the order of `list_revisions.sh`'s output does.

aj-stein-nist replied:
https://github.com/usnistgov/OSCAL-Reference/pull/23/commits/2a7ced9018dd1a6bc0a40415aa4792ae55a2a682

nikitawootten-nist commented on .github/workflows/pages.yaml:
```diff
@@ -46,7 +59,12 @@ jobs:
         # GHA runners have 2 CPUs
-        run: make site -j2
+        if: env.REVISIONS == 'unset'
+        run: make site -j2 PROTOTYPE_BRANCHES_PREFIX="${{ env.PROTOTYPE_BRANCHES_PREFIX }}"
```
B, maybe nitpicky: To keep things a bit more readable, could we try folding flags into one:
```suggestion
        run: |
          make site -j2 \
            ${{ github.event.inputs.revisions != '' && format("REVISIONS='{0}'", github.event.inputs.revisions) || '' }} \
            ${{ github.event.inputs.revisions == '' && github.event.inputs.prefix && format("PROTOTYPE_BRANCHES_PREFIX='{0}'", github.event.inputs.prefix) || '' }}
```

This has the major benefit of making the default value still controlled by the Makefile. If adopted the `env` block could be removed.

aj-stein-nist replied:
I am happy with that change. **EDIT:** sorry I now see there might be an issue with that approach, I will message you.

aj-stein-nist replied:
Adjusted per feedback in https://github.com/usnistgov/OSCAL-Reference/pull/23/commits/2a7ced9018dd1a6bc0a40415aa4792ae55a2a682.

nikitawootten-nist reviewed (changes requested):
Some requested changes regarding the repurposing of `list_revisions.sh` and the workflow logic

aj-stein-nist commented:
> @nikitawootten-nist, I am not sure we would want to or not want to adjust the cache keys based on tags as opposed to accompanying revisions, that would make this maybe more complicated, or do I just have to find a way of calling the [updated block in the relevant GitHub Actions block](https://github.com/usnistgov/OSCAL-Reference/blob/d5eac0ba934c301cc537f35d7c5065ffcea83d5e/.github/workflows/pages.yaml#L32)?

On Monday, @nikitawootten-nist we should make sure we understand how this will work and its caching is logical.

@Arminta-Jenkins-NIST it changed a bit so we should re-review on Monday morning.

aj-stein-nist commented:
Examples from my fork with `main` aligned to the PR branch for testing "once merged" after Nikita's feedback changes by me were completed.

1. [Running as is with no alternate revisions or branch prefix](https://github.com/aj-stein-nist/OSCAL-Reference/actions/runs/6279027875)
2. [Running with targeted revisions, skipping one of the two prototype branches](https://github.com/aj-stein-nist/OSCAL-Reference/actions/runs/6279726735) in the revisions argument set to `develop v1.0.0 v1.0.1 v1.0.2 v1.0.3 v1.0.4 v1.0.5 v1.0.6 v1.1.0 v1.1.1 prototype-mapping-model`, excluding `prototype-shared-responsibility-model` from the list so it is "unpublished" on next run)
3. [Running with no revision change but picking a new custom prefix](https://github.com/aj-stein-nist/OSCAL-Reference/actions/runs/6279728061), `does_not_exist`, which will return nothing and unpublish all branches from before with prefix `prototype`

It seems 3 needs to be investigated as the prototype prefix branch still includes the prototype-mapping branch from 2 even though it should not, so maybe this is not a done deal so to speak.

aj-stein-nist committed:
Rebase branch selection changes for #19 on updated main.

aj-stein-nist committed:
Check for cache and restore of branches on #19.

When testing on my fork, I had forgotten my fork did not have any of the
prototype branches and for the use case of using GHA on the upstream,
we only check "origin" (the one remote) for now, but that was not as
intended during test that must happen off fork. This testing reminded me
that caching should use the branches as an output and only restore from
cache if they are defined, especially if the default prototype prefix is
used, but nothing is returned, so don't cache empty.

aj-stein-nist committed:
Correct branch prefix quoting for #19.

aj-stein-nist committed:
Switch to Makefile approach for #19.

nikitawootten-nist commented on .github/workflows/pages.yaml:
```diff
@@ -63,8 +71,18 @@ jobs:
         id: get-revisions
         run: |
-          echo revisions_hash=$(cd support/OSCAL; git tag | grep -E '^v\d+\.\d+\.\d+$' | sha256sum) >> $GITHUB_OUTPUT
+          echo revisions_hash=$(make list-tags | sha256sum) >> $GITHUB_OUTPUT
```
```suggestion
          echo revisions_hash=$(make --silent list-tags | sha256sum) >> $GITHUB_OUTPUT
```

aj-stein-nist replied:
Fixed in pairing.

nikitawootten-nist commented on .github/workflows/pages.yaml:
```diff
@@ -63,8 +71,18 @@ jobs:
+        id: get-prototype-branches
+        run: |
+          # Wildcards after a stable tag prefix like v* for v1.0.0 will not work, so generate outputs for content and data paths for next step
+          echo branches="$(make list-prototype-branches $PREFIX)" >> $GITHUB_OUTPUT
```
```suggestion
          echo branches="$(make --silent list-prototype-branches PROTOTYPE_BRANCHES_PREFIX='$PREFIX')" >> $GITHUB_OUTPUT
```

aj-stein-nist replied:
Fixed in pairing.

nikitawootten-nist commented on support/list_branches.sh:
```diff
@@ -0,0 +1,29 @@
+ROOT_DIR="${SCRIPT_DIR}/.."
+OSCAL_DIR="${ROOT_DIR}/support/OSCAL"
+
+[[ -z "${1-}" ]] && { echo "Error: Necessary git remote not set"; usage; exit 1; }
+[[ -z "${2-}" ]] && { echo "Error: Necessary git branch prefix not set"; usage; exit 1; }
```
lines must be moved **below** the definition of `usage()`

aj-stein-nist replied:
Fixed in pairing.

aj-stein-nist committed:
Pairing advice, do not use restore keys for #19

Per recommendation from @nikitawootten-nist do not use restore keys for
the cache as it uses lax matching on sub-components of the hash (support
files in this workflow) when the specific content is missed. For the
prototype branches, this is counterproductive and ill-advised.

aj-stein-nist commented:
OK @nikitawootten-nist, ready for reviewed based on fork testing this appears to be good, and we confirmed my impressions about missed branch deletions with incorrect. We can finalize beginning of next sprint if necessary.

aj-stein-nist committed:
Exempt link checking on prototype branches for #19.

aj-stein-nist commented:
OK per discussion in sprint planning and pairing this afternoon, we shall exempt the prototype branches from link checking with the expectation once you try to point into OSCAL `develop`, that is the step beyond prototype and you need to check those links before it can get published. That commit has been added on the (now properly) rebased PR branch below.

https://github.com/usnistgov/OSCAL-Reference/pull/23/commits/98405dc857b97de26f08a73d1a6799f726728a9a

aj-stein-nist commented:
Still seem to be getting some failures and trying to clean them up, for others I am testing locally with `make linkcheck LYCHEE_EXTRA_FLAGS="--exclude-path=*/models/prototype*" REVISIONS="develop $(make -s list-prototype-branches)"` to mimic the implied expansion of shell commands from the Makefile and GHA defaults.

aj-stein-nist commented:
I will need to finish examining later, but seems that somehow when testing locally the bad `mapping` links I have from building develop and prototype branches is in develop, not the prototype branches. 🤷

https://gist.github.com/aj-stein-nist/8a1ae4ab17f57ef9ffcb6022248e9439

I will make clean for the third or fourth time after rebasing with Nikita on the call, but ... odd?

aj-stein-nist commented:
Since I've been reading logs a lot, something I need to consider fixing in this PR or in a follow-up one.

```sh
[WARN ] WARNING: `--exclude-file` is deprecated and will soon be removed; use `.lycheeignore` file to ignore URL patterns instead. To exclude paths of files and directories, use `--exclude-path`.
```

aj-stein-nist committed:
Stop regex filter, exclude linkcheck paths with printf for #19.

aj-stein-nist commented:
OK, @nikitawootten-nist and others, it turns out that the only thing breaking now is timeouts on a itu.int link that has been working in my browser but even failed intermittently with links and such. I resolved the previous issue by getting rid of the regex wildcard technique for use with `lychee` to ignore paths to prototype branches (and not develop) for links that are not prime-time ready yet. So what's below didn't work.

```sh
lychee --exclude-file ./support/lychee_ignore.txt  \
  --config ./support/lychee.toml \
  --output lychee_report.md \
  --verbose --format markdown \
  --exclude-path="*public/models*" \
  site/public/**/*.html
```
I tried many, many different paths, then even specific paths with the wildcards before the path, after it, and both, with `public/develop` or `develop` and it still reported bad links checks there, so something isn't right there. Before we were using GHA workflows to stub in the branch name between wildcards, but I blew that way, when I realized some of the errors I was adding on purpose could not be silenced in the key places, so this approach was a DOA.

So I pivoted to using the list of branches to build a list of `--exclude-path=site/public/models/branchname1 --exclude-path=site/public/models/branchname2` to the extra arguments override in the Makefile.

```sh
lychee --exclude-file ./support/lychee_ignore.txt  \
  --config ./support/lychee.toml \
  --output lychee_report.md \
  --verbose --format markdown \
  --exclude-path=site/public/models/prototype-mapping-model \
  --exclude-path=site/public/models/prototype-shared-responsibility-model \
  site/public/**/*.html
```
Then we use the GitHub Actions workflow to pass in the orchestrated by the pages.yaml workflow by using the Linux/Unix utility `printf` to take that list of branches and create the stringified form of that one branch name at a time. The `\055` is the encoding of the first `-` in `--` for `--exclude-path` because `printf` sees that as an argument and you must be careful about that, hence the Stackoverflow note. The relevant argument for others to understand what I mean is `linkcheck_exclude_path_arg=$(...)`.

```yaml
# Truncated other irrelevant code
      - name: Get the list of tagged revisions (for cache)
        id: get-revisions
        run: |
          echo revisions_hash=$(make --silent list-tags | sha256sum) >> $GITHUB_OUTPUT
        shell: bash
      - name: Get the list of prototype branches (for cache)
        id: get-prototype-branches
        run: |
          {
            # Wildcards after a stable tag prefix like v* for v1.0.0 will not work, so generate outputs for content and data paths for next step
            branches="$(make --silent list-prototype-branches PROTOTYPE_BRANCHES_PREFIX=$PREFIX)"
            echo branches="$branches"
            [[ -n "${branches-}" ]] && echo content_paths="$(printf "site/content/models/%s " $branches)"
            [[ -n "${branches-}" ]] && echo data_paths="$(printf "site/content/data/%s " $branches)"
            # Be forewarned, printf -- has a special meaning so we must encode the first - as \055 in -- to avoid errors
            # Explanation: https://unix.stackexchange.com/a/658388
            [[ -n "${branches-}" ]] && echo linkcheck_exclude_path_arg=$(printf "\055-exclude-path='site/public/models/%s' "  $branches)
            # Prototype branch content may change so hash branch name and current commit hash for its HEAD as restore key part
            echo branches_hash="$(cd "${OSCAL_DIR}"; git ls-remote origin $PREFIX | sha256sum | cut -d ' ' -f 1)"
          } >> $GITHUB_OUTPUT
# Truncated other irrelevant code
      - name: Link Check
        id: linkcheck
        run: |
          make linkcheck LYCHEE_EXTRA_FLAGS='--github-token ${{ secrets.GITHUB_TOKEN }} \
            ${{ steps.get-prototype-branches.outputs.linkcheck_exclude_path_arg && format('{0}', steps.get-prototype-branches.outputs.linkcheck_exclude_path_arg) || '' }}'
```
I will keep re-running, but I think it works for now. I am working on a bug report to the lychee folk, but this is ready for review and merge I think.

aj-stein-nist committed:
Remove deprecated `.lycheeignore` approach for #19

As indicated in the comment below and the warning provided, we need to
migrate the configuration to the forward-looking configuration not the
older `.lycheeignore` approach.

https://github.com/usnistgov/OSCAL-Reference/pull/23#issuecomment-1745425997

nikitawootten-nist commented on .github/workflows/pages.yaml:
```diff
@@ -63,8 +71,24 @@ jobs:
+            [[ -n "${branches-}" ]] && echo data_paths="$(printf "site/content/data/%s " $branches)"
+            # Be forewarned, printf -- has a special meaning so we must encode the first - as \055 in -- to avoid errors
+            # Explanation: https://unix.stackexchange.com/a/658388
+            [[ -n "${branches-}" ]] && echo linkcheck_exclude_path_arg=$(printf "\055-exclude-path='site/public/models/%s' "  $branches)
```
NB: That sounds like it was painful to debug, nice catch

aj-stein-nist replied:
Well, I got through it! 😅

nikitawootten-nist commented on Makefile:
```diff
@@ -17,9 +17,11 @@ clean: clean-modeldoc clean-site clean-release-assets clean-linkcheck ## Clean a
 
+PROTOTYPE_BRANCHES_REMOTE:=origin
+PROTOTYPE_BRANCHES_PREFIX:=prototype
 # Override REVISIONS to build a subset of the site or a special branch
 #   (e.g. `make site REVISIONS='v1.1.0 my-special-branch'`)
```
NB: Suggest adding more context for future maintainers here:
```suggestion
# `REVISIONS` contains a list of Git refs (specifically branches and tags) of the OSCAL repository that documentation will be built for.
# By default, `REVISIONS` includes:
# 1. The `develop` branch, showcasing changes staged for the next release.
# 2. Tagged releases matching a semver pattern (e.g. `v.1.0.0`).
# 3. Prototype branches matching the specified `PROTOTYPE_BRANCHES_REMOTE` and `PROTOTYPE_BRANCHES_PREFIX`.
# For local development, `REVISIONS` can be overridden to build a subset of the site
#   (e.g. `make site REVISIONS='v1.1.0 my-special-branch'`)
```

nikitawootten-nist commented on support/list_branches.sh:
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+# List all branches that generated documentation should target
```
NB: Suggest clarifying the purpose of this script
```suggestion
# List all OSCAL branches from the specified git remote matching the specified prefix (e.g. prototype)
```

nikitawootten-nist reviewed (approved):
Minor NB doc suggestions below, looks good!

aj-stein-nist committed:
Address doc string clarification feedbacks from @nikitawootten-nist.

Co-authored-by: Nikita Wootten <nikita.wootten@nist.gov>
